libshortfin/setup.py

```diff
@@ -55,8 +70,127 @@ def copy_extensions_to_source(self, *args, **kwargs):
+    # Ideally, CMake would let us reconfigure this dynamically... but it does
+    # not (and gets very confused).
+    PYTHON_STAMP_FILE = os.path.join(cmake_build_dir, "python_stamp.txt")
+    if os.path.exists(PYTHON_STAMP_FILE):
+        with open(PYTHON_STAMP_FILE, "rt") as f:
+            actual_stamp_contents = f.read()
+            if actual_stamp_contents == expected_stamp_contents:
+                # All good.
+                return
+
+    # Mismatch or not found. Clean it.
+    cmake_cache_file = os.path.join(cmake_build_dir, "CMakeCache.txt")
+    if os.path.exists(cmake_cache_file):
+        print("Removing CMakeCache.txt because Python version changed", file=sys.stderr)
+        os.remove(cmake_cache_file)
+
+    # And write.
+    with open(PYTHON_STAMP_FILE, "wt") as f:
+        f.write(expected_stamp_contents)
+
+
+class CMakeBuildPy(_build_py):
+    def run(self):
+        # The super-class handles the pure python build.
+        super().run()
+
+        # Build using cmake if not in prebuild mode.
+        if not is_cpp_prebuilt():
+
+            # Build extension using cmake.
+            print("*****************************", file=sys.stderr)
+            print("* Building libshortfin      *", file=sys.stderr)
+            print("*****************************", file=sys.stderr)
+
+            cfg = os.getenv("SHORTFIN_CMAKE_BUILD_TYPE", "Release")
+
+            CMAKE_BUILD_DIR = BINARY_DIR
+
+            # Configure CMake.
+            os.makedirs(BINARY_DIR, exist_ok=True)
+            maybe_nuke_cmake_cache(CMAKE_BUILD_DIR)
+            print(f"CMake build dir: {CMAKE_BUILD_DIR}", file=sys.stderr)
+            cmake_args = [
+                "-GNinja",
+                "--log-level=VERBOSE",
+                "-DSHORTFIN_BUNDLE_DEPS=ON",
+                f"-DCMAKE_BUILD_TYPE={cfg}",
+                "-DSHORTFIN_BUILD_PYTHON_BINDINGS=ON",
+                # TODO: This shouldn't be hardcoded... but shortfin doesn't
+                # compile without it.
+                "-DCMAKE_C_COMPILER=clang",
+                "-DCMAKE_CXX_COMPILER=clang++",
+            ]
+
+            # Only do a from-scratch configure if not already configured.
+            cmake_cache_file = os.path.join(CMAKE_BUILD_DIR, "CMakeCache.txt")
+            if not os.path.exists(cmake_cache_file):
+                print(f"Configuring with: {cmake_args}", file=sys.stderr)
+                subprocess.check_call(
+                    ["cmake", SOURCE_DIR] + cmake_args, cwd=CMAKE_BUILD_DIR
+                )
+            else:
+                print(f"Not re-configing (already configured)", file=sys.stderr)
+
+            # Build.
+            subprocess.check_call(["cmake", "--build", "."], cwd=CMAKE_BUILD_DIR)
+            print("Build complete.", file=sys.stderr)
+
+            # We only take _shortfin_default from the build.
+            target_dir = os.path.join(
+                os.path.abspath(self.build_lib), "_shortfin_default"
```

Comment:
Fine to leave for now but we're going to end up with multiple cmake invocations producing different variants of this directory (ie for tracing builds, etc). For context, see how iree runtime/setup.py does it.

Reply:
Yeah i used runtime/setup.py as reference. I think it should be easy to do that, I would add a suffix here for the specific build, something like _shortfin_default_tracy.